File: core/src/main/java/uk/lobsterdoodle/namepicker/storage/StorageUseCase.java
package uk.lobsterdoodle.namepicker.storage;

import org.greenrobot.eventbus.Subscribe;

import java.util.List;

import javax.inject.Inject;

import uk.lobsterdoodle.namepicker.addgroup.AddNameToGroupEvent;
import uk.lobsterdoodle.namepicker.creategroup.CreateGroupDetailsEvent;
import uk.lobsterdoodle.namepicker.events.EventBus;
import uk.lobsterdoodle.namepicker.model.Group;
import uk.lobsterdoodle.namepicker.model.GroupDetails;
import uk.lobsterdoodle.namepicker.model.Name;
import uk.lobsterdoodle.namepicker.namelist.RetrieveGroupNamesEvent;
import uk.lobsterdoodle.namepicker.overview.OverviewBecameVisibleEvent;
import uk.lobsterdoodle.namepicker.overview.OverviewCardCellData;
import uk.lobsterdoodle.namepicker.overview.OverviewRetrievedEvent;
import uk.lobsterdoodle.namepicker.selection.NameStateChangedEvent;

import static com.google.common.collect.Lists.transform;

public class StorageUseCase {
    private final KeyValueStore storage;
    private final RemoteDb remoteDb;
    private final DbHelper db;
    private final EventBus bus;

    @Inject
    public StorageUseCase(KeyValueStore storage, RemoteDb remoteDb, DbHelper db, EventBus bus) {
        this.storage = storage;
        this.remoteDb = remoteDb;
        this.db = db;
        this.bus = bus;
        bus.register(this);
    }

    @Subscribe
    public void on(OverviewBecameVisibleEvent event) {
        final List<Group> groupList = db.getAllGroups();
        final List<OverviewCardCellData> cellData = transform(groupList,
                group -> new OverviewCardCellData(group.getDetails().getId(), group.getDetails().getName(), group.getNames().size()));

        bus.post(new OverviewRetrievedEvent(cellData));
        for (OverviewCardCellData item : cellData)
            remoteDb.editGroupDetails(item.getGroupId(), item.getListTitle(), item.getNameCount());
    }

    @Subscribe
    public void on(CreateGroupDetailsEvent event) {
        final long groupId = db.createGroup(event.getGroupName());
        bus.post(new GroupCreationSuccessfulEvent(groupId, event.getGroupName()));
    }

    @Subscribe
    public void on(AddNameToGroupEvent event) {
        db.addNameToGroup(event.getGroupId(), event.getName());
        final List<Name> groupNames = db.retrieveGroupNames(event.getGroupId());
        bus.post(new GroupNamesRetrievedEvent(event.getGroupId(), groupNames));
        bus.post(new NameAddedSuccessfullyEvent());
    }

    @Subscribe
    public void on(RetrieveGroupNamesEvent event) {
        final List<Name> retrieveGroupNames = db.retrieveGroupNames(event.getGroupId());
        bus.post(new GroupNamesRetrievedEvent(event.getGroupId(), retrieveGroupNames));
    }

    @Subscribe
    public void on(DeleteNameEvent event) {
        final Name deletedName = db.removeName(event.getNameId());
        bus.post(new NameDeletedSuccessfullyEvent(deletedName.getName()));
    }

    @Subscribe
    public void on(DeleteGroupEvent event) {
        final GroupDetails details = db.removeGroup(event.getGroupId());
        bus.post(new GroupDeletedSuccessfullyEvent(details.getName()));
    }

    @Subscribe
    public void on(EditGroupDetailsEvent event) {
        db.editGroupName(event.getGroupId(), event.getGroupName());
        bus.post(new GroupNameEditedSuccessfullyEvent());
    }

    @Subscribe
    public void on(RetrieveGroupDetailsEvent event) {
        final GroupDetails details = db.retrieveGroupDetails(event.getGroupId());
        final List<Name> names = db.retrieveGroupNames(event.getGroupId());
        bus.post(new GroupDetailsRetrievedSuccessfullyEvent(details));
        remoteDb.editGroupDetails(event.getGroupId(), details.getName(), names.size());
    }

    @Subscribe
    public void on(NameStateChangedEvent event) {
        db.updateName(event.getName());
    }

    @Subscribe
    public void on(MassNameStateChangedEvent event) {
        db.toggleAllNamesInGroup(event.getGroupId(), event.getToggleOn());
    }
}

File: core/src/main/kotlin/uk/lobsterdoodle/namepicker/storage/StorageUseCase.kt
package uk.lobsterdoodle.namepicker.storage

import org.greenrobot.eventbus.Subscribe
import uk.lobsterdoodle.namepicker.addgroup.AddNameToGroupEvent
import uk.lobsterdoodle.namepicker.creategroup.CreateGroupDetailsEvent
import uk.lobsterdoodle.namepicker.events.EventBus
import uk.lobsterdoodle.namepicker.namelist.RetrieveGroupNamesEvent
import uk.lobsterdoodle.namepicker.overview.OverviewBecameVisibleEvent
import uk.lobsterdoodle.namepicker.overview.OverviewCardCellData
import uk.lobsterdoodle.namepicker.overview.OverviewRetrievedEvent
import uk.lobsterdoodle.namepicker.selection.NameStateChangedEvent
import javax.inject.Inject

class StorageUseCase @Inject
constructor(private val storage: KeyValueStore, private val remoteDb: RemoteDb, private val db: DbHelper, private val bus: EventBus) {

    init {
        bus.register(this)
    }

    @Subscribe
    fun on(event: OverviewBecameVisibleEvent) {
        val groupList = db.allGroups
        val cellData = groupList.map { (details, names) -> OverviewCardCellData(details.id, details.name, names.size) }

        bus.post(OverviewRetrievedEvent(cellData))
        for ((groupId, listTitle, nameCount) in cellData)
            remoteDb.editGroupDetails(groupId, listTitle, nameCount)
    }

    @Subscribe
    fun on(event: CreateGroupDetailsEvent) {
        val groupId = db.createGroup(event.groupName)
        bus.post(GroupCreationSuccessfulEvent(groupId, event.groupName))
    }

    @Subscribe
    fun on(event: AddNameToGroupEvent) {
        db.addNameToGroup(event.groupId, event.name)
        val groupNames = db.retrieveGroupNames(event.groupId)
        bus.post(GroupNamesRetrievedEvent(event.groupId, groupNames))
        bus.post(NameAddedSuccessfullyEvent)
    }

    @Subscribe
    fun on(event: RetrieveGroupNamesEvent) {
        val retrieveGroupNames = db.retrieveGroupNames(event.groupId)
        bus.post(GroupNamesRetrievedEvent(event.groupId, retrieveGroupNames))
    }

    @Subscribe
    fun on(event: DeleteNameEvent) {
        val name = db.removeName(event.nameId)
        name?.let { bus.post(NameDeletedSuccessfullyEvent(name.name)) }
    }

    @Subscribe
    fun on(event: DeleteGroupEvent) {
        val name = db.removeGroup(event.groupId)
        name?.let { bus.post(GroupDeletedSuccessfullyEvent(name.name)) }
    }

    @Subscribe
    fun on(event: EditGroupDetailsEvent) {
        db.editGroupName(event.groupId, event.groupName)
        bus.post(GroupNameEditedSuccessfullyEvent)
    }

    @Subscribe
    fun on(event: RetrieveGroupDetailsEvent) {
        val details = db.retrieveGroupDetails(event.groupId)
        val names = db.retrieveGroupNames(event.groupId)
        details?.let {
            bus.post(GroupDetailsRetrievedSuccessfullyEvent(details))
            remoteDb.editGroupDetails(event.groupId, details.name, names.size)
        }
    }

    @Subscribe
    fun on(event: NameStateChangedEvent) {
        db.updateName(event.name)
    }

    @Subscribe
    fun on(event: MassNameStateChangedEvent) {
        db.toggleAllNamesInGroup(event.groupId, event.toggleOn)
    }
}
